### DISCUSSION TITLE
Can we provide ACLs through webhooks?

### DISCUSSION THREAD

**Discussion by markandpathak** [vernemq/vernemq#1993] [Q&A]
Consider following situation. We have number of clients connecting to the broker. Each client should publish message to only certain allowed topics which may change over time and often calculated on the fly. Now as mentioned [here](https://docs.vernemq.com/configuring-vernemq/db-auth#introduction-and-general-setup), ACL can be specified through a database and is cached till the connection persists.

Is there any way, this ACL JSON data is provided through webhook modifiers in response for like `auth_on_register` ? If so I could not find any documentation on that. Any guidance is appreciated.

And if this is not possible wouldn't it be a great feature to add ACL through a web-hook modifier, for example in response of `auth_on_register` as given below:
```
{
    "result": "ok",
    "modifiers": {
        "max_message_size": 65535,
        "max_inflight_messages": 10000,
        "retry_interval": 20000
        "acl": {
            "pattern": "a/+/c",
            "max_qos": 2,
            "max_payload_size": 128,
            "allowed_retain": true
        }
    }
}
```
We can give dynamically generated ACL each time hook is called for a new connection. And as mentioned the ACL is anyways cached till the connection persists, so no further web-hook calls are required for each publish.

 

Similar strategy can be implemented with `auth_on_publish` webhooks. But issue is this method may induce serious traffic to the webhook provider when number of messages and client is high as each message needs to be authorized to get published. 

Let me know if any further information is required.

**Comment by ioolkos** [comment]
@markandpathak thanks for opening this discussion!
Isn't this mainly about shifting work between Plugin and Broker, though? As it is, you would do the authorization decision in the WebHooks plugin, where you have some sort of ACL handling anyway. 
Do you need something I don't see?

**Reply by markandpathak** [reply]
Thank you for the reply @ioolkos !

Yes, indeed handling will be required. Sorry if I put the question in wrong way. The end goal is to dynamically and efficiently control ACL for all clients through web-hooks. 

`auth_on_register` handles the auth on connection level, `auth_on_publish` handles the auth on message level. But what I am suggesting is that **there should be a level in webhook which allows auth on topic level** which directly defines ACL of topics for a prticular client. 

Let me make it clear by a example: In Auth using files section you can specify an ACL entry
```
# ACL for user 'john'
user john
topic write data_from_john
```
Above entry allows me to get a topic level control that user `john` can publish to topic `data_from_john`. **But there is no straight forward way to do same thing using webhooks.** 

Somehow if we still like to implement it with webhooks, we need to implement `auth_on_publish` endpoint which is called on each published message, the endpoint checks whether client is authorized for publishing to topic or not. But checking webhook endpoint for each message is very inefficient, Instead it should happen locally in the broker once endpoint provides ACL of topics. 

A possible way to apply ACL to a client connection is to provide ACL through `auth_on_register` webhook. If client is authorized to connect, webhook endpoint **returns result ok along wiht an ACL of topics for that client connection**. Then whenever client publishes to the topic, same ACL is used locally (just like it happens with file based ACLs) on the broker itself resulting in efficient access control through web-hooks.
